scm/mock.go

```diff
@@ -0,0 +1,408 @@
+package scm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/quickfeed/quickfeed/qf"
+)
+
+var testOrgs = []*qf.Organization{
+	{
+		ID:   1,
+		Name: "qfTestOrg",
+	},
+	{
+		ID:   2,
+		Name: "DAT320",
+	},
+	{
+		ID:   3,
+		Name: "DATx20-2019",
+	},
+	{
+		ID:   4,
+		Name: "DATx20-2020",
+	},
+}
+
+// MockSCM implements the SCM interface.
+// TODO(meling) many of the methods below are not implemented.
+type MockSCM struct {
+	Repositories  map[uint64]*Repository
+	Organizations map[uint64]*qf.Organization
+	Hooks         map[uint64]*Hook
+	Teams         map[uint64]*Team
+}
+
+// NewMockSCMClient returns a new mock client implementing the SCM interface.
+func NewMockSCMClient() *MockSCM {
+	s := &MockSCM{
+		Repositories:  make(map[uint64]*Repository),
+		Organizations: make(map[uint64]*qf.Organization),
+		Hooks:         make(map[uint64]*Hook),
+		Teams:         make(map[uint64]*Team),
+	}
+	s.initOrganizations()
+	return s
+}
+
+func (MockSCM) Clone(_ context.Context, opt *CloneOptions) (string, error) {
+	cloneDir := filepath.Join(opt.DestDir, repoDir(opt))
+	// This is a hack to make sure the lab1 directory exists,
+	// required by the web/rebuild_test.go:TestRebuildSubmissions()
+	lab1Dir := filepath.Join(cloneDir, "lab1")
+	err := os.MkdirAll(lab1Dir, 0o700)
+	if err != nil {
+		return "", err
+	}
+	return cloneDir, err
+}
+
+// UpdateOrganization implements the SCM interface.
+func (s *MockSCM) UpdateOrganization(ctx context.Context, opt *OrganizationOptions) error {
+	if !opt.valid() {
+		return fmt.Errorf("invalid argument: %+v", opt)
+	}
+	if _, err := s.GetOrganization(ctx, &GetOrgOptions{Name: opt.Name}); err != nil {
+		return errors.New("organization not found")
+	}
+	return nil
+}
+
+// GetOrganization implements the SCM interface.
+func (s *MockSCM) GetOrganization(_ context.Context, opt *GetOrgOptions) (*qf.Organization, error) {
+	if !opt.valid() {
+		return nil, fmt.Errorf("invalid argument: %+v", opt)
+	}
+	if opt.ID < 1 {
+		for _, org := range s.Organizations {
+			if org.Name == opt.Name {
+				return org, nil
+			}
+		}
+	}
+	org, ok := s.Organizations[opt.ID]
+	if !ok {
+		return nil, errors.New("organization not found")
+	}
+	return org, nil
+}
+
+// CreateRepository implements the SCM interface.
+func (s *MockSCM) CreateRepository(ctx context.Context, opt *CreateRepositoryOptions) (*Repository, error) {
+	if !opt.valid() {
+		return nil, fmt.Errorf("invalid argument: %+v", opt)
+	}
+	org, err := s.GetOrganization(ctx, &GetOrgOptions{
+		ID:   opt.Organization.ID,
+		Name: opt.Organization.Name,
+	})
+	if err != nil {
+		return nil, err
+	}
+	repo := &Repository{
+		ID:      uint64(len(s.Repositories) + 1),
```

Comment:
Assigning ID like this might end up overwriting another repository:

```go
// s.Repositories
{
    1: repo1
    2: repo2
    3: repo3
}
// Delete repo2
{
    1: repo1
    3: repo3
}
// Create repo, len(s.Repositories) + 1 == 3
{
    1: repo1
    3: repo4
}

Reply:
In the frontend I ended up with something similar to this for generating IDs:

```go
var IDs = make(map[string]uint64, 0)

func nextID(key string) uint64 {
	if _, ok := IDs[key]; !ok {
		IDs[key] = 1
	} else {
		IDs[key]++
	}
	return IDs[key]
}

nextID("repo") // 1
nextID("repo") // 2
nextID("teams") // 1
// Delete a repository 
nextID("repo") // 3
```

Reply:
Agree, even if it worked without issues all these years it can still cause unexpected problems for new tests. Same for `Hooks` and `Teams`.

Reply:
Added a safe ID generator, thank you for the inspiration.